File: 2226-maximum-candies-allocated-to-k-children/2226-maximum-candies-allocated-to-k-children.cpp
class Solution {
public:
    int maximumCandies(vector<int>& candies, long long k) {
        long long sum = 0;
        for (auto candy: candies) sum += candy;
        if (sum < k) return 0;
        
        long long low = 1, high = *max_element(candies.begin(), candies.end());
        int res;
        
        while (low <= high){
            long long mid = low + (high - low) / 2;
            
            if (helper(mid, candies, k)){
                res = mid;
                low = mid + 1;
            }else high = mid - 1;
        }
        return res;
    }
private:
    bool helper(long long pile_size, vector<int>& candies, long long k){
        long long count = 0;
        for (auto candy: candies){
            count += candy / pile_size;
        }
        return count >= k;
    }
};

File: 2226-maximum-candies-allocated-to-k-children/2226-maximum-candies-allocated-to-k-children.py
class Solution:
    def maximumCandies(self, candies: List[int], k: int) -> int:
        low, high = 1, max(candies)
        
        if sum(candies) < k: return 0 
        
        def find_part(pile_size):
            count = 0
            for candy in candies:
                count += candy // pile_size
            return count >= k
        
        while low <= high:
            mid = low + (high - low) // 2
            if find_part(mid):
                res = mid
                low = mid + 1
            else:
                high = mid - 1
                
        return res
        
        